### PR TITLE
[GFC-901][DC] welsh translation 

### PR DESCRIPTION
**Pull request by DarylMConway**
[GFC-711][DC] adding lang custom class

[GFC-711][DC] adding lang in to tests

[GFC-711][DC] adding lang in to tests

[GFC-901][DC] works for components, working on sections

[GFC-901][DC] works for sections

[GFC-901][DC] works for choices

[GFC-901][DC] works for auto-language detection

[GFC-901][DC] rebased on to master

### CODE REVIEW

File: app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala [hunk 1]
```diff
@@ -253,8 +255,8 @@ class FormComponentMaker(json: JsValue) {
...
         case (IsOrientation(HorizontalOrientation), Some(x :: xs), IsMultivalue(MultivalueYes), Selections(selections), oHelpText) => Choice(Checkbox, NonEmptyList(x, xs),          Horizontal, selections, oHelpText).asRight
         case (IsOrientation(HorizontalOrientation), Some(x :: xs), IsMultivalue(MultivalueNo),  Selections(selections), oHelpText) => Choice(Radio,    NonEmptyList(x, xs),          Horizontal, selections, oHelpText).asRight
-        case (IsOrientation(YesNoOrientation),      None,          IsMultivalue(MultivalueNo),  Selections(selections), oHelpText) => Choice(YesNo,    NonEmptyList.of("Yes", "No"), Horizontal, selections, oHelpText).asRight
-        case (IsOrientation(YesNoOrientation),      _,             _,                           Selections(selections), oHelpText) => Choice(YesNo,    NonEmptyList.of("Yes", "No"), Horizontal, selections, oHelpText).asRight
+        case (IsOrientation(YesNoOrientation),      None,          IsMultivalue(MultivalueNo),  Selections(selections), oHelpText) => Choice(YesNo,    NonEmptyList.of(toLocalsiedString("Yes"), toLocalsiedString("No")), Horizontal, selections, oHelpText).asRight
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala]: "Yes" and "No" aren't localised. Welsh alternatives are missing.

File: app/uk/gov/hmrc/gform/formtemplate/SectionHelper.scala [hunk 2]
```diff
@@ -78,18 +82,21 @@ object SectionHelper {
...
-      if (field.label.contains("$n") || (field.shortName.isDefined && field.shortName.get.contains("$n"))) {
+      {
         field.copy(
-          label = buildRepeatingText(Some(field.label), 1).get,
-          shortName = buildRepeatingText(field.shortName, 1))
-      } else {
-        field
+          label = LocalisedString(labelRepeatingGroupComponents(field.label.m, 1)),
+          shortName =
+            if (field.shortName.isEmpty) None
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/SectionHelper.scala]: This `if ... else` should be replaced by `field.shortName.map(...)`

File: app/uk/gov/hmrc/gform/formtemplate/SectionHelper.scala [hunk 3]
```diff
@@ -65,8 +67,10 @@ object SectionHelper {
                 List(
                   fieldInGroup.copy(
                     id = FormComponentId(fieldName),
-                    label = buildRepeatingText(Some(fieldInGroup.label), i + 1).getOrElse(""),
-                    shortName = buildRepeatingText(fieldInGroup.shortName, i + 1)
+                    label = LocalisedString(labelRepeatingGroupComponents(fieldInGroup.label.m, i + 1)),
+                    shortName =
+                      if (fieldInGroup.shortName.isEmpty) None
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/SectionHelper.scala]: This `if ... else` should be replaced by `field.shortName.map(...)` also.

File: app/uk/gov/hmrc/gform/sharedmodel/LangADT.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  def langADTToString(langADT: LangADT): String = langADT.langADTToString
+
+  def stringToLangADT(string: String): LangADT = string match {
+    case "cy" => Cy
+    case _    => En
+  }
+
+  private def convertToLang(jsValue: JsValue): JsResult[LangADT] = {
+    println("jsValue " + jsValue)
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/LangADT.scala]: Remove `println`.

File: test/uk/gov/hmrc/gform/sharedmodel/FormatValidationSpec.scala [hunk 5]
```diff
@@ -81,11 +81,12 @@ class FormatValidationSpec extends Spec {
 
   def createTest(format: String, constraint: TextConstraint) = {
     val json: JsValue = makeJson(format)
+    println(json)
```
Comment (josef-vlach) [inline test/uk/gov/hmrc/gform/sharedmodel/FormatValidationSpec.scala]: Remove `println`

File: app/uk/gov/hmrc/gform/sharedmodel/LocalisedString.scala [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+package uk.gov.hmrc.gform.sharedmodel
+
+import play.api.libs.json.{ Format, Reads }
+import play.api.libs.functional.syntax._
+import uk.gov.hmrc.gform.formtemplate.LangMapMaker
+import uk.gov.hmrc.gform.sharedmodel.formtemplate.JsonUtils
+
+case class LocalisedString(m: Map[LangADT, String]) {
+  def value(implicit l: LangADT): String = m.getOrElse(l, m.getOrElse(LangADT.En, ""))
+  def values: List[String] = m.values.toList
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/LocalisedString.scala]: Are these `values` needed?

File: app/uk/gov/hmrc/gform/sharedmodel/LocalisedString.scala [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+  val basicFormat: Format[LocalisedString] =
+    formatMap.inmap[LocalisedString](m => LocalisedString(m), ls => ls.m)
+
+  val templateReads: Reads[LocalisedString] = LangMapMaker.langADTMapReads.map(LocalisedString.apply)
+
+  implicit val format: Format[LocalisedString] = FormatWithTemplateReadFallback2(basicFormat, templateReads)
+}
+
+object FormatWithTemplateReadFallback2 {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/LocalisedString.scala]: Name this object properly.

File: app/uk/gov/hmrc/gform/formtemplate/LangMapMaker.scala [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    override def pure[A](x: A): JsResult[A] = JsSuccess(x)
+    override def ap[A, B](ff: JsResult[A => B])(fa: JsResult[A]): JsResult[B] =
+      for {
+        aToB <- ff
+        a    <- fa
+      } yield aToB(a)
+  }
+
+  private def convertToLangMap(jsValue: JsValue): JsResult[Map[LangADT, String]] =
+    jsValue match {
```
[the remark this answers is not in the document]
Comment (DarylMConway) [reply app/uk/gov/hmrc/gform/formtemplate/LangMapMaker.scala]: added "case other => JsError("Expected string, or language pair" + other)"

**Pull request closed without merging**